packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/subscribe.ts

```diff
@@ -0,0 +1,294 @@
+    } else {
+      this._eventsQueue = this.createQueues('Events', this.props?.eventsQueueProps, this.props?.eventsDeadLetterQueueProps);
+    }
+    this.environment[`${this.parentService.id.toUpperCase()}_EVENTS_QUEUE_URL`] = this._eventsQueue.queueUrl;
+
+    if (this.props?.topicSubscriptions) {
+      const eventsQueueSubscriptions = [];
+      for (const topicSubscription of this.props.topicSubscriptions) {
+        const topic = topicSubscription.topic;
+        const topicName = topic.node.id;
+
+        if (topicSubscription.queue && topicSubscription.queueProps) {
+          throw Error(`You can only specify either a queue or queue props for creating a queue for topic ${topicName}.`);
+        }
+
+        if (!topicSubscription.queue && !topicSubscription.queueProps && !topicSubscription.deadLetterQueueProps) {
+          topic.addSubscription(new subscription.SqsSubscription(this._eventsQueue));
+          eventsQueueSubscriptions.push(topic);
+        } else {
+          const topicSubscriptionQueue = topicSubscription.queue ??
+            this.createQueues(topicName, topicSubscription.queueProps, topicSubscription.deadLetterQueueProps);
+          this.environment[`${this.parentService.id.toUpperCase()}_${topicName.toUpperCase()}_QUEUE_URL`] = topicSubscriptionQueue.queueUrl;
+
+          topic.addSubscription(new subscription.SqsSubscription(topicSubscriptionQueue));
+          this.subscriptions?.push({
+            topics: [topic],
+            subscriptionQueue: topicSubscriptionQueue,
+          });
+        }
+      }
+      this.subscriptions?.push({
+        topics: eventsQueueSubscriptions,
+        subscriptionQueue: this._eventsQueue,
+      });
+    }
+  }
+
+  /**
+   * Add hooks to the main application extension so that it is modified to
+   * add the queue URLs to the container environment.
+   */
+  public addHooks() {
+    const container = this.parentService.serviceDescription.get('service-container') as Container;
+
+    if (!container) {
+      throw new Error('Subscribe extension requires an application extension');
+    }
+
+    container.addContainerMutatingHook(new SubscribeMutatingHook({
+      environment: this.environment,
+    }));
+  }
+
+  /**
+   * After the task definition has been created, this hook grants SQS permissions to the task role.
+   *
+   * @param taskDefinition The created task definition
+   */
+  public useTaskDefinition(taskDefinition: ecs.TaskDefinition) {
+    if (this.subscriptions) {
+      for (const subs of this.subscriptions) {
+        subs.subscriptionQueue.grantConsumeMessages(taskDefinition.taskRole);
+      }
+    } else {
+      this._eventsQueue.grantConsumeMessages(taskDefinition.taskRole);
+    }
+
+    for (const deadLetterQueue of this.deadLetterQueues) {
+      deadLetterQueue.grantConsumeMessages(taskDefinition.taskRole);
+    }
+  }
+
+  private createQueues(name: string, queueProps?: sqs.QueueProps, deadLetterQueueProps?: DeadLetterQueueProps) : sqs.IQueue {
```

Comment:
It might be possible that some folks will not want a dead letter queue. For example if a queue has highly valuable messages going to it, and a message is failing to be processed, they may want to leave it on the queue and fix the code of their queue consuming service, rather than rejecting the message to a dead letter queue. We should make sure dead letter queue is optional